### PR TITLE
[ENG-1385] Use diamond contract for sale contract

### PR DESCRIPTION
**Pull request by deveshrawat**
Description:

1. Update ERC721Sale and REC1155Sale methods
2. Update examples
3. Add functionality to check for tags to support auction and sale 
4. Update version

### CODE REVIEW

File: examples/listForSale.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ async function main() {
       // Use and parse existing NFTs to put for sale
       const nft = refinable.createNft({
         type: parameters[2] as TokenType,
-        chainId: Chain.BscTestnet,
```
Comment (maximgeerinck) [inline examples/listForSale.ts:37]: example should stay testnet

File: src/config/sdk.ts [hunk 2]
```diff
@@ -16,14 +16,14 @@ export const contractsTags = {
...
-    ContractTag.SaleV3_0_0,
+    ContractTag.SaleV4_0_0,
     ContractTag.AuctionV3_1_0,
     ContractTag.SaleNonceHolderV1_0_0,
     ContractTag.TransferProxyV1_0_0,
     ContractTag.AirdropV1_0_0,
   ],
   [Environment.Local]: [
-    ContractTag.SaleV3_0_0,
+    ContractTag.SaleV4_0_0,
```
Comment (maximgeerinck) [inline src/config/sdk.ts:26]: we have to find a way to only do  this for the imx chain

File: src/nft/AbstractNFT.ts [hunk 3]
```diff
@@ -439,6 +427,10 @@ export abstract class AbstractNFT {
     serviceFeeUserAddress: string,
     address: string
   ): Promise<number> {
+    const diamondSale = await this.refinable.contracts.getRefinableContract(
```
Comment (maximgeerinck) [inline src/nft/AbstractNFT.ts:430]: this won't always be a diamondSale, we will do this in phases.

**Review by maximgeerinck — CHANGES_REQUESTED** [review]
as it stands right now i don't understand it :D 

We should only use the diamond contract for the chain we'll introduce next. the other chains should still use the sale contracts as defined before.

the diamond contract also acts as a big proxy towards the sales, so from my understanding, it should be sufficient to put this is a the value of `saleContract` ? not sure if i'm missing something.

File: examples/listForSale.ts (hunk 1 above)
Comment (deveshrawat) [reply examples/listForSale.ts:37]: will update this.

**Comment by deveshrawat** [comment]
Putting value in saleContract's place is not enough as proxies also needs taken care, as we don't have proxies for diamond contracts.

**Comment by maximgeerinck** [comment]
> Putting value in saleContract's place is not enough as proxies also needs taken care, as we don't have proxies for diamond contracts.

still we should have backsupport for chains without diamond :)

File: src/config/sdk.ts (hunk 2 above)
Comment (deveshrawat) [reply src/config/sdk.ts:26]: That was not the idea initially, we weren't just wanted to use diamond Contract for our existing solution? Need to think about it if we want to use this for specific chain >>>

File: src/nft/AbstractNFT.ts (hunk 3 above)
Comment (deveshrawat) [reply src/nft/AbstractNFT.ts]: Yes indeed this name is misleading changed the variable name

File: src/config/sdk.ts (hunk 2 above)
Comment (deveshrawat) [reply src/config/sdk.ts]: Added a check for  a particular chain

File: src/config/sdk.ts [hunk 4]
```diff
@@ -1,32 +1,48 @@
...
+      ];
+    case Environment.Testnet:
+      return [
+        chain === ChainType.BSC
+          ? ContractTag.SaleV4_0_0
+          : ContractTag.SaleV3_0_1,
+        ContractTag.AuctionV3_1_0,
+        ContractTag.SaleNonceHolderV1_0_0,
+        ContractTag.TransferProxyV1_0_0,
+        ContractTag.AirdropV1_0_0,
```
Comment (maximgeerinck) [inline src/config/sdk.ts:35]: isn't auction also v4?

File: tests/unit-tests/erc1155.test.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  PriceCurrency,
+  Refinable,
+  StandardRoyaltyStrategy,
+} from "../../src";
+
+describe("Refinable", () => {
+  let refinable: Refinable;
+  const PRIVATE_KEY = process.env.PRIVATE_KEY as string;
+  const API_KEY = process.env.API_KEY as string;
+  const wallet = initializeWallet(PRIVATE_KEY, Chain.Local);
```
Comment (maximgeerinck) [inline tests/unit-tests/erc1155.test.ts:17]: Better to work with a mock of wallet

File: tests/unit-tests/erc1155.test.ts [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  beforeAll(async () => {
+    refinable = await Refinable.create(wallet, API_KEY, {
+      waitConfirmations: 1,
+      environment: Environment.Local,
+    });
```
Comment (maximgeerinck) [inline tests/unit-tests/erc1155.test.ts:22]: you need your local chain to work now?

**Review by maximgeerinck — APPROVED** [review]
The tests look more like integration tests and will require you to have a local chain running.
Honestly, this is hard to automate as local chains on CI/CD is a bit tricky for now.

File: src/config/sdk.ts (hunk 4 above)
Comment (deveshrawat) [reply src/config/sdk.ts:35]: We are testing it for Sale contracts for now but if everything goes well we will update the auction contracts as well, this is the beauty of diamond standard we can add a single contract while keeping everything as before, so currently we are deploying only sale contracts.

File: tests/unit-tests/erc1155.test.ts (hunk 5 above)
Comment (deveshrawat) [reply tests/unit-tests/erc1155.test.ts:17]: Yeah, that sounds a better Idea will update this

**Pull request merged**